[LLDB] Add more helper functions to CompilerType class (second try).

cmtice opened the pull request:
This adds 23 new helper functions to LLDB's CompilerType class, things like IsSmartPtrType, IsPromotableIntegerType,
GetNumberofNonEmptyBaseClasses, and GetTemplateArgumentType (to name a few).

It also has run clang-format on the files CompilerType.{h,cpp}.

These helper functions are needed as part of the implementation for the Data Inspection Language, (see
https://discourse.llvm.org/t/rfc-data-inspection-language/69893).

cmtice committed:
[LLDB] Add more helper functions to CompilerType class (second try).

This adds 23 new helper functions to LLDB's CompilerType class, things
like IsSmartPtrType, IsPromotableIntegerType,
GetNumberofNonEmptyBaseClasses, and GetTemplateArgumentType (to name a
few).

It also has run clang-format on the files CompilerType.{h,cpp}.

These helper functions are needed as part of the implementation for
the Data Inspection Language, (see
https://discourse.llvm.org/t/rfc-data-inspection-language/69893).

bulbazord commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -302,6 +303,256 @@ bool CompilerType::IsBeingDefined() const {
+  if (name == canonical_name) {
+    return llvm::formatv("'{0}'", name);
+  }
+  return llvm::formatv("'{0}' (aka '{1}')", name, canonical_name);
```
Suggestion: Instead of "'foo' (aka 'bar')" maybe something like

"'foo' (canonically referred to as 'bar')"?

My reasoning is 'aka' does not specify where this information is "known" or who knows it. Maybe canonical is also the wrong word but I think it would be useful to tell folks more than just "this type is also known as this other type". Maybe it's just me though.

cmtice replied:
Done.

bulbazord reviewed (commented):
Seems okay to me in general. Left some inline comments and I also agree with all of Jonas's inline comments I saw too.

felipepiovezan commented on lldb/include/lldb/Symbol/CompilerType.h:
```diff
@@ -436,8 +482,8 @@ class CompilerType {
 
   /// Dump to stdout.
-  void DumpTypeDescription(lldb::DescriptionLevel level =
-                           lldb::eDescriptionLevelFull) const;
```
I understand the desire to fix formatting issues, but let's avoid doing them in a big patch like this. If things needs to be reverted, as this patch was, such a formatting change should not need to be reverted.

This can also cause unnecessary work for downstream forks in case of merge conflicts (the first patch, the revert, the second patch, etc).

cmtice replied:
Personally I agree with you. GIthub pull request creation did not give me any choice: It would not allow me to create a PR that did not pass clang-format. If you know of some way to bypass that requirement, PLEASE tell me!!

DavidSpickett replied:
> It would not allow me to create a PR that did not pass clang-format.

Can you be more specific here, you could not open a PR at all without it being formatted? If that's the case it doesn't seem like what anyone would have intended.

Your other workaround here would be to do a PR to reformat these files, then rebase this onto that after landing it. Review has already started though, so maybe do that as the last thing after approvals.

cmtice replied:
DavidSpickett:  I created the PR, sort of, but... if you look at the bottom of this PR (for example) there's a small section that says "All Checks Passed". If you open that, you will see that one of the checks is a clang-format check.  My original PR did not pass that check, so github would not "finish" creating the PR, i.e. it would not allow me to ask for reviews on it.

DavidSpickett replied:
That's annoying for sure. I'm a member of the repo, and I was able to add reviewers to this test PR https://github.com/llvm/llvm-project/pull/7381 that failed clang-format. Then again, Github tells me you're also a member, so that can't be the problem. Strange.

If you can figure out the sequence of events or reproduce it, please open an issue because I'm not sure it's intended behaviour. Perhaps we're trying to be strict but still, there should be some way to ignore the check when needed.

felipepiovezan replied:
I could be wrong here, but AFAICT the bot should _only_ check for formatting changes inside the diff of the commit, not inside the entire file. If this is not what the bot is doing, it is something we can probably improve upon.

In the meantime though, my suggestion is to manually inspect the bot output: if it is complaining about lines not touched by this commit, just ignore that. This is favorable to editing lines that are not related to the patch

felipepiovezan replied:
I see this comment in [another PR](https://github.com/llvm/llvm-project/pull/73252#issuecomment-1832242278)

> NB: not applying all the clang-format recommendations as it affects lines I'm not editing, and I don't want to pollute git-blame

That makes me think the bot is not running clang-format-diff? https://clang.llvm.org/docs/ClangFormat.html#script-for-patch-reformatting

cmtice replied:
Question:  For this PR, do you want me to un-do the clang-format changes that don't relate to my changes?

felipepiovezan replied:
Yes, please :)

cmtice replied:
Done.

felipepiovezan commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -54,7 +54,7 @@ bool CompilerType::IsArrayType(CompilerType *element_type_ptr, uint64_t *size,
     if (auto type_system_sp = GetTypeSystem())
       return type_system_sp->IsArrayType(m_type, element_type_ptr, size,
-                                      is_incomplete);
+                                         is_incomplete);
```
same issue with formatting as mentioned before

cmtice replied:
Done.

felipepiovezan commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -157,7 +157,8 @@ bool CompilerType::IsBlockPointerType(
   if (IsValid())
     if (auto type_system_sp = GetTypeSystem())
-      return type_system_sp->IsBlockPointerType(m_type, function_pointer_type_ptr);
+      return type_system_sp->IsBlockPointerType(m_type,
```
same issue with formatting as mentioned before

cmtice replied:
Done.

felipepiovezan commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -249,7 +250,7 @@ bool CompilerType::IsPossibleDynamicType(CompilerType *dynamic_pointee_type,
     if (auto type_system_sp = GetTypeSystem())
       return type_system_sp->IsPossibleDynamicType(m_type, dynamic_pointee_type,
-                                                check_cplusplus, check_objc);
+                                                   check_cplusplus, check_objc);
```
same issue with formatting as mentioned before

cmtice replied:
Done.

felipepiovezan commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -302,6 +303,256 @@ bool CompilerType::IsBeingDefined() const {
+      "^std::unique_ptr<.+>(( )?&)?$");
+  static llvm::Regex k_libcxx_std_shared_ptr_regex_2(
+      "^std::shared_ptr<.+>(( )?&)?$");
+  static llvm::Regex k_libcxx_std_weak_ptr_regex_2(
```
Let's make these `const`s

cmtice replied:
Done.

felipepiovezan commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -302,6 +303,256 @@ bool CompilerType::IsBeingDefined() const {
+
+bool CompilerType::IsInteger() const {
+  // This is used when you don't care about the signedness of the integer.
+  bool is_signed;
```
+1

cmtice replied:
Done.

felipepiovezan commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -302,6 +303,256 @@ bool CompilerType::IsBeingDefined() const {
+
+  default:
+    return false;
+  }
```
Rationale: https://llvm.org/docs/CodingStandards.html#don-t-use-default-labels-in-fully-covered-switches-over-enumerations

cmtice replied:
Done.

cmtice replied:
Done.

felipepiovezan commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -302,6 +303,256 @@ bool CompilerType::IsBeingDefined() const {
+bool CompilerType::IsVirtualBase(CompilerType target_base,
+                                 CompilerType *virtual_base,
+                                 bool carry_virtual) const {
+  if (CompareTypes(target_base)) {
+    return carry_virtual;
+  }
```
There are many other instances of this in the code below

cmtice replied:
Done (everywhere, I think).

felipepiovezan commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -302,6 +303,256 @@ bool CompilerType::IsBeingDefined() const {
+          lldb::eTypeClassUnion);
+}
+
+// Checks whether `target_base` is a virtual base of `type` (direct or
+// indirect). If it is, stores the first virtual base type on the path from
```
+1

felipepiovezan replied:
This comment should also explain what `virtual_base` and `carry_virtual` are

cmtice replied:
Done (mostly). I still need to check on 'carry_virtual' and add a comment about it.

adrian-prantl commented on lldb/include/lldb/Symbol/CompilerType.h:
```diff
@@ -194,6 +192,54 @@ class CompilerType {
 
   bool IsVoidType() const;
+
+  bool IsSmartPtrType() const;
```
Some methods like this one could benefit from a Doxygen comment that explains what the semantics are.

cmtice replied:
Done.

adrian-prantl commented on lldb/include/lldb/Symbol/CompilerType.h:
```diff
@@ -194,6 +192,54 @@ class CompilerType {
+  bool IsVirtualBase(CompilerType target_base, CompilerType *virtual_base,
+                     bool carry_virtual = false) const;
+
+  bool IsContextuallyConvertibleToBool() const;
```
This seems extremely specific to the C language family. I don't think that's a problem, but it would be nice to at least document that this may be only defined in TypeSystemClang?

cmtice replied:
Done? I think?

cmtice replied:
Really done. :-)

adrian-prantl commented on lldb/include/lldb/Symbol/CompilerType.h:
```diff
@@ -194,6 +192,54 @@ class CompilerType {
+
+  CompilerType GetTemplateArgumentType(uint32_t idx);
+
+  CompilerType GetSmartPtrPointeeType();
```
There seems to be some overlap between this and the C++ shared_ptr synthetic child provider. I think it would be good if either the formatter uses this functionality or if the intended user of this function uses the synthetic children instead.

cmtice replied:
Could you elaborate on this a little bit more please. I'm still learning the LLDB type system, and I don't quite understand what you're asking me to do.

adrian-prantl replied:
Here you are adding a function that takes a CompilerType holding an STL  smart pointer and ask for the pointee type. We also have synthetic child formatters used by `frame variable` that allow evaluating `*my_shared_ptr`. From that I was assuming that there must be some overlap between what the dataformatter provides and what this function does.
What I don't know is whether synthetic children provided by a formatter are a good fit for what you are trying to do, but maybe the inverse is true, and the existing data formatter should be implemented in terms of this new GetSmartPtrPointeeType() API.
I just want to make sure we're not introducing two places in LLDB that have to know about STL smart pointers while covering slightly different functionality.

jimingham replied:
Note that this function returns the type of the shared pointer, whereas the synthetic child providers return values of that type.  So if anything you would implement the formatter on top of this type recognition (though that only gets you a little way there, you also have to find where the value is in the smart pointer.)

It's a little awkward that these API's speak of "smart pointer" when they really mean "the system library provided memory managing smart pointer".  But I can't think of a good name to make this explicit.

However, that brings us round to the synthetic child support for "smart pointers".  That is a feature that's not limited to the standard library's notion of a smart pointer.  It really just tells lldb how to interpret the `->` in:

(lldb) frame var foo->bar

The implementation of this feature doesn't require any knowledge of what the arrow operator for foo does (it is not even required to have one).  For instance, you could use it to have lldb log uses of the `->` operator for this type in the expression parser, or whatever.  The only requirement is that you define a synthetic child provider that emulates the -> operator, returning a ValueObject from the appropriate API in the provider.

So at some point someone could redo the STL smart pointer comprehension in lldb to use the synthetic child arrow operator emulation, that's a much broader feature that what this API is promising.

cmtice replied:
Do you want me to try to update the data formatter to you my function in this PR, or should we commit this PR without that change, with the promise that I will submit a follow-up PR with the data formatter change?

Jim, I follow what you're saying but I'm not sure if there's a concrete action you're asking me to take?

adrian-prantl replied:
> Do you want me to try to update the data formatter to you my function in this PR, or should we commit this PR without that change, with the promise that I will submit a follow-up PR with the data formatter change?

If it's easy to do, sure — but I'm also fine with postponing it for a follow-up commit. I just want to make sure we have a clear plan for how to make that happen.

cmtice replied:
I would prefer to do it in a follow-up commit. (My next PR!).

jimingham replied:
I was mostly trying to say that the ValueObject support for "faking the arrow operator effects" is orthogonal to having the type system contain the notion of "smart pointers".  

I also think the whole naming of this is confusing.  What is a "smart pointer"?  The IsSmartPtr states some intent about this:

`  /// This determines if the type is a shared, unique or weak pointer, either
  /// from stdlibc++ or libc+++.
  bool IsSmartPtrType() const;
`

But that is also a little odd.  This is a general type system query but doesn't in fact have anything to do with the type system.  Rather with a particular implementation in the C++ STL - which is orthogonal to the C++ type system...  So it seems like an awfully C++ specific API to have in the Typesystem.

jimingham replied:
Having a comment saying "This is defined only for C++ types" in a function from CompilerType.h seems wrong.

cmtice replied:
I can remove these helper functions from the CompilerType class and put them into my parser class for DIL instead, if you think that would be better.

cmtice replied:
Clarification: I meant "I can remove the helper functions related to smart pointers."

cmtice commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -302,6 +303,256 @@ bool CompilerType::IsBeingDefined() const {
+  // These regular expressions cover shared, unique and weak pointers both from
+  // stdlibc++ and libc+++.
+
+  static llvm::Regex k_libcxx_std_unique_ptr_regex(
```
It indicates the that these are constants.  I can remove it if you like?

bulbazord replied:
Oh I see, I don't mind it being there. Just hadn't seen it before. If they're constants though, then please mark them `const` as well.

cmtice replied:
Done.

cmtice commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -302,6 +303,256 @@ bool CompilerType::IsBeingDefined() const {
+  auto name = GetTypeName();
+  auto canonical_name = GetCanonicalType().GetTypeName();
+  if (name.IsEmpty() || canonical_name.IsEmpty()) {
+    return "''"; // should not happen
```
I believe this will only happen for broken input; I'll update the comment to make that clear.

cmtice replied:
Done.

cmtice committed:
[LLDB] Add more helper functions to CompilerType class (second try).

Address most of the reviewer comments:
  - Fix formatting issues
  - Add/Move Doxygen function comments in header file.
  - Undo clang-format changes to code that I didn't touch.
  - Other requested fixes.

cmtice commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -302,6 +303,256 @@ bool CompilerType::IsBeingDefined() const {
+      "^std::__[[:alnum:]]+::shared_ptr<.+>(( )?&)?$");
+  static llvm::Regex k_libcxx_std_weak_ptr_regex(
+      "^std::__[[:alnum:]]+::weak_ptr<.+>(( )?&)?$");
+  //
```
Done.

cmtice commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -302,6 +303,256 @@ bool CompilerType::IsBeingDefined() const {
+      "^std::shared_ptr<.+>(( )?&)?$");
+  static llvm::Regex k_libcxx_std_weak_ptr_regex_2(
+      "^std::weak_ptr<.+>(( )?&)?$");
+  //
```
Done.

cmtice commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -302,6 +303,256 @@ bool CompilerType::IsBeingDefined() const {
+}
+
+bool CompilerType::IsInteger() const {
+  // This is used when you don't care about the signedness of the integer.
```
Done.

cmtice commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -302,6 +303,256 @@ bool CompilerType::IsBeingDefined() const {
+bool CompilerType::IsEnumerationIntegerTypeSigned() const {
+  if (IsValid()) {
+    return GetEnumerationIntegerType().GetTypeInfo() & lldb::eTypeIsSigned;
+  }
```
Done (for entire change).

cmtice commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -302,6 +303,256 @@ bool CompilerType::IsBeingDefined() const {
+    case lldb::eTypeClassEnumeration: return "enum";
+    case lldb::eTypeClassStruct:      return "struct";
+    case lldb::eTypeClassUnion:       return "union";
+    // clang-format on
```
Fixed the formatting & removed the comments.

cmtice commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -302,6 +303,256 @@ bool CompilerType::IsBeingDefined() const {
 
+bool CompilerType::IsSmartPtrType() const {
+  // These regular expressions cover shared, unique and weak pointers both from
+  // stdlibc++ and libc+++.
```
Done.  Renamed the libstdc++ variable names to contain 'libstdc++'  instead of 'libc++'.

cmtice commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -302,6 +303,256 @@ bool CompilerType::IsBeingDefined() const {
+}
+
+CompilerType CompilerType::GetSmartPtrPointeeType() {
+  assert(IsSmartPtrType() &&
```
Fixed. Instead of asserting, if it's not a smart pointer I call GetPointeeType.

cmtice committed:
[LLDB] Add more helper functions to CompilerType class (second try).

Update comment for "IsVirtualBase" member function as requested.

cmtice commented:
I believe i have addressed all the review comments. Please let me know if there's anything else I need to do and/or (hopefully) approve this so I can merge it. 

Thanks!

jimingham commented:
They really do seem odd in the general CompilerType...

Jim


> On Dec 5, 2023, at 4:35 PM, cmtice ***@***.***> wrote:
> 
> 
> @cmtice commented on this pull request.
> 
> In lldb/include/lldb/Symbol/CompilerType.h <https://github.com/llvm/llvm-project/pull/73472#discussion_r1416471222>:
> 
> > +
> +  bool IsContextuallyConvertibleToBool() const;
> +
> +  bool IsBasicType() const;
> +
> +  std::string TypeDescription();
> +
> +  bool CompareTypes(CompilerType rhs) const;
> +
> +  const char *GetTypeTag();
> +
> +  uint32_t GetNumberOfNonEmptyBaseClasses();
> +
> +  CompilerType GetTemplateArgumentType(uint32_t idx);
> +
> +  CompilerType GetSmartPtrPointeeType();
> Clarification: I meant "I can remove the helper functions related to smart pointers."
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/llvm/llvm-project/pull/73472#discussion_r1416471222>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVW4JAG5WJJFUNJSL2HLYH64WZAVCNFSM6AAAAAA73I4PWOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTONRWGM3TAOJXGY>.
> You are receiving this because you are on a team that was mentioned.
>

cmtice committed:
[LLDB] Add more helper functions to CompilerType class (second try).

Remove smart pointer functions from the new helper functions for
CompilerType class.

cmtice commented:
I've removed the Smart pointer functions from this PR.

Are there any other changes that I need to make? If not, could someone approve this?

bulbazord reviewed (commented):
Seems okay to me... Quick question: Why do you need `CompilerType::GetTemplateArgumentType`? This seems pretty specific to C++. Is there a more general concept we could capture? Like something related to "Generics"?

cmtice commented:
@bulbazord Actually, I think you're right. I use that in my smart pointer type code.  I will pull that out as well (I overlooked it before).

cmtice committed:
[LLDB] Add more helper functions to CompilerType class (second try).

Remove GetTemplateArgumentType from list of new helper functions (it's
too C++-specific).

cmtice commented:
Ok, I have now removed CompilerType::GetTemplateArgumentType from the new helper functions.

bulbazord reviewed (approved):
Seems okay to me. I don't see anything in here that wouldn't make sense in at least some other languages (other than C++).

adrian-prantl commented on lldb/source/Symbol/CompilerType.cpp:
```diff
@@ -302,6 +302,195 @@ bool CompilerType::IsBeingDefined() const {
+  if (IsEnumerationType())
+    return IsEnumerationIntegerTypeSigned();
+
+  return GetTypeInfo() & lldb::eTypeIsSigned;
```
This function looks like it's working around a bug in GetTypeInfo?

adrian-prantl replied:
(i.e., could this be fixed by modifying TypeSystemClang::GetTypeInfo to return the sign bit on enums?

cmtice replied:
@adrian-prantl Actually I think it already works properly without special-casing enums (I went back and looked and could not figure out why I had done this). I've cleaned it up now and removed the special casing.  Are you OK with me committing this now?

cmtice committed:
[LLDB] Add more helper functions to CompilerType class (second try).

Clean up implementation of IsSigned function.

adrian-prantl reviewed (approved):
Thanks for addressing all my feedback!
